src/classes/CW_AudienceGroupTriggerHandlerTest.cls

```diff
@@ -0,0 +1,236 @@
+/**
+* File Name   :    CW_AudienceGroupTriggerHandlerTest.cls
+* Description :    D-106803o
+*                 
+*
+* Modification Log
+* ============================================================================
+* Ver Date       Author         Modification
+* --- ---------- -------------- --------------------------
+* 0.1 24/04/2014 Alex  Panka		Initial
+ */
+ 
+@isTest
+public with sharing class CW_AudienceGroupTriggerHandlerTest {
+	
+	static testMethod void ValidationRuleTest() {
+		
+		CW_Campaign__c camp = TestDataUtil.createCampaign();
+      insert camp;
+      //We muset use single call for testDataUtil for uniq names
+      List<CW_Audience_Group__c> listAllGrps  =TestDataUtil.createAudienceGroups(camp.Id,6);
+      
+		List<CW_Audience_Group__c> listControlGrps  = new List <CW_Audience_Group__c>();
+		List<CW_Audience_Group__c> listControlGrps2 = new List <CW_Audience_Group__c>();
+		List<CW_Audience_Group__c> listNormalGrps   = new List <CW_Audience_Group__c>();
+		listControlGrps.Add(listAllGrps.get(0));
+		listControlGrps.Add(listAllGrps.get(1));
+		listControlGrps2.Add(listAllGrps.get(2));
+		listControlGrps2.Add(listAllGrps.get(3));
+		listNormalGrps.Add(listAllGrps.get(4));
+		listNormalGrps.Add(listAllGrps.get(5));
+		
+		insert listControlGrps;
+		insert listControlGrps2;
+		
+
+
+		for(integer i=0;i<2;i++){
+			listNormalGrps.get(i).Is_Control_Group__c=false;
+			listNormalGrps.get(i).Control_Group__c=listControlGrps.get(i).id;
+		}
+		insert listNormalGrps;
+	   //Test if we can chage witout ref
+	   for(CW_Audience_Group__c Rec :listControlGrps2){
+	   	Rec.Is_Control_Group__c=false;
+	   }
+	   update listControlGrps2;
+	   for(CW_Audience_Group__c Rec :listControlGrps2){
+	   	Rec.Is_Control_Group__c=true;
+	   }
+		update listControlGrps2;
+	   
+	   try{
+	   	insert listAllGrps;
+	   	System.assert(false,'Validation of Unique Name on Audience control group not rise error');
+	   }
+	   catch(Exception E){}	 
+	   
+	   
+	}
+	
+	static testMethod void controlGroupTest()
+	{
+		CW_Campaign__c camp = TestDataUtil.createCampaign();
+      	insert camp;
+        
+		CW_Audience_Group__c controlGroup = TestDataUtil.createAudienceGroup(camp.Id);
+		controlGroup.Name = 'Control1';
+		controlGroup.Is_Control_Group__c = true;
+		insert controlGroup;
+		CW_Audience_Group__c targetGroup = TestDataUtil.createAudienceGroup(camp.Id);
+		targetGroup.Name = 'Target1';
+		targetGroup.Is_Control_Group__c = false;
+		targetGroup.Control_Group__c = controlGroup.Id;
+		insert targetGroup;
+		
+		Test.startTest();
+		targetGroup.Is_Control_Group__c = true;
+		update targetGroup;
+		Test.stopTest();
+		
+		CW_Audience_Group__c reselectTarget = [select Is_Control_Group__c, Control_Group__c 
```

Comment:
Problem: Unbound SOQL query. Reason: a SOQL query can return up to 50,000 records. If this number is exceeded a runtime exception is thrown. Suggestion: cap the max number of records returned adding a LIMIT clause to the query.